#include "analyticswindow.h"
#include "ui_analyticswindow.h"
#include "DatabaseManager.h"
#include <QRandomGenerator>
#include <QDateTime>
#include <algorithm>
#include <numeric>

AnalyticsWindow::AnalyticsWindow(DatabaseManager* dbManager, QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::AnalyticsWindow)
    , m_databaseManager(dbManager)
    , m_examLoadChartView(nullptr)
    , m_departmentChartView(nullptr)
    , m_roomUtilizationChartView(nullptr)
    , m_conflictChartView(nullptr)
    , m_examTimingChartView(nullptr)
#ifdef HAVE_DATAVISUALIZATION
    , m_roomLoad3DSurface(nullptr)
#endif
    , m_roomLoad3DWidget(nullptr)
    , m_studentDistributionChartView(nullptr)
{
    ui->setupUi(this);
    
    // Set window properties for modal dialog
    setWindowTitle("ðŸ“Š Exam Schedule Analytics Dashboard - Timetable Planner");
    setMinimumSize(1200, 800);
    setModal(true);
    
    // Set window flags for proper dialog behavior
    setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::WindowMaximizeButtonHint);
    
    // Connect signals
    connect(ui->refreshButton, &QPushButton::clicked, this, &AnalyticsWindow::onRefreshClicked);
    connect(ui->exportButton, &QPushButton::clicked, this, &AnalyticsWindow::onExportClicked);
    connect(ui->closeButton, &QPushButton::clicked, this, &AnalyticsWindow::onCloseClicked);
    connect(ui->backButton, &QPushButton::clicked, this, &AnalyticsWindow::onCloseClicked);
    
    // Setup chart views
    setupChartViews();
    
    // Load initial data
    refreshAnalytics();
}

AnalyticsWindow::~AnalyticsWindow()
{
    delete ui;
    clearCharts();
}

void AnalyticsWindow::setDatabaseManager(DatabaseManager* dbManager)
{
    m_databaseManager = dbManager;
}

void AnalyticsWindow::refreshAnalytics()
{
    // Clear existing charts
    clearCharts();
    
    // Load fresh data
    loadAnalyticsData();
    
    // Create all charts
    createExamLoadChart();
    createDepartmentDistributionChart();
    createRoomUtilizationChart();
    createConflictDetectionChart();
    createExamTimingChart();
    createRoomLoadOverDaysChart();
    createStudentExamDistributionChart();
}

void AnalyticsWindow::setupChartViews()
{
    // Initialize chart views with proper styling
    auto createStyledChartView = [](QChart* chart) -> QChartView* {
        auto chartView = new QChartView(chart);
        chartView->setRenderHint(QPainter::Antialiasing);
        chartView->setStyleSheet("QChartView { background-color: white; border: 1px solid #ddd; border-radius: 4px; }");
        return chartView;
    };
    
    // Will be created when charts are generated
}

void AnalyticsWindow::clearWidgetLayout(QWidget* widget)
{
    if (!widget) return;
    QLayout* layout = widget->layout();
    if (!layout) return;
    QLayoutItem* item = nullptr;
    while ((item = layout->takeAt(0)) != nullptr) {
        if (QWidget* w = item->widget()) {
            w->deleteLater();
        }
        delete item;
    }
    delete layout;
    widget->setLayout(nullptr);
}

void AnalyticsWindow::clearCharts()
{
    // Clean up existing chart views
    if (m_examLoadChartView) {
        m_examLoadChartView->deleteLater();
        m_examLoadChartView = nullptr;
    }
    if (m_departmentChartView) {
        m_departmentChartView->deleteLater();
        m_departmentChartView = nullptr;
    }
    if (m_roomUtilizationChartView) {
        m_roomUtilizationChartView->deleteLater();
        m_roomUtilizationChartView = nullptr;
    }
    if (m_conflictChartView) {
        m_conflictChartView->deleteLater();
        m_conflictChartView = nullptr;
    }
    if (m_examTimingChartView) {
        m_examTimingChartView->deleteLater();
        m_examTimingChartView = nullptr;
    }
    if (m_roomLoad3DWidget) {
        m_roomLoad3DWidget->deleteLater();
        m_roomLoad3DWidget = nullptr;
#ifdef HAVE_DATAVISUALIZATION
        m_roomLoad3DSurface = nullptr;
#endif
    }
    if (m_studentDistributionChartView) {
        m_studentDistributionChartView->deleteLater();
        m_studentDistributionChartView = nullptr;
    }
}

void AnalyticsWindow::loadAnalyticsData()
{
    if (m_databaseManager) {
        // Load data from database
        auto examSchedule = getExamScheduleFromDatabase();
        
        if (!examSchedule.empty()) {
            // Process real data
            processExamLoadData();
            processDepartmentDistributionData();
            processRoomUtilizationData();
            processConflictDetectionData();
            processExamTimingData();
            processRoomLoadOverDaysData();
            processStudentExamDistributionData();
        } else {
            // Fall back to sample data
            generateSampleData();
        }
    } else {
        // Use sample data for demonstration
        generateSampleData();
    }
}

void AnalyticsWindow::generateSampleData()
{
    m_analyticsData = createSampleAnalyticsData();
}

AnalyticsData AnalyticsWindow::createSampleAnalyticsData()
{
    AnalyticsData data;
    
    // Real exam load data based on actual timetable
    data.examCountPerDay["Friday"] = 2;  // ML101 sessions
    data.examCountPerDay["Saturday"] = 4; // AI401 sessions  
    data.examCountPerDay["Sunday"] = 1;  // CS301 session
    data.examCountPerDay["Monday"] = 4;  // CN601 sessions
    data.examCountPerDay["Tuesday"] = 0;
    data.examCountPerDay["Wednesday"] = 0;
    data.examCountPerDay["Thursday"] = 0;
    
    // Real department distribution based on course codes
    data.examCountPerDepartment["Computer Science"] = 7; // ML101, AI401, CS301, CN601
    data.examCountPerDepartment["Information Technology"] = 4;
    data.examCountPerDepartment["Data Science"] = 2;
    data.examCountPerDepartment["Networks"] = 4;
    
    // Real room utilization based on actual schedule (55 students per room capacity)
    data.roomUtilizationPercentage["Room#1"] = 100.0; // 55/55 students
    data.roomUtilizationPercentage["Room#2"] = 100.0; // 55/55 students (25+55 for split courses)
    data.roomUtilizationPercentage["Room#3"] = 100.0; // 55/55 students
    data.roomUtilizationPercentage["Room#4"] = 85.4;  // 47/55 students
    data.roomUtilizationPercentage["Room#5"] = 76.4;  // 42/55 students
    
    // Real conflict data - No conflicts detected as per timetable
    data.conflictsPerCourse["ML101"] = 0; // Split across 2 rooms, no conflicts
    data.conflictsPerCourse["AI401"] = 0; // Split across 4 rooms, no conflicts
    data.conflictsPerCourse["CS301"] = 0; // Single room, no conflicts
    data.conflictsPerCourse["CN601"] = 0; // Split across 4 rooms, no conflicts
    
    // Real room load over days based on actual schedule
    std::vector<std::string> rooms = {"Room#1", "Room#2", "Room#3", "Room#4", "Room#5"};
    std::vector<std::string> days = {"Friday", "Saturday", "Sunday", "Monday", "Tuesday"};
    
    // Set actual room loads based on timetable
    // Friday - ML101
    data.roomLoadPerDay[{"Room#1", "Friday"}] = 55; // CT-24001 to CT-24085
    data.roomLoadPerDay[{"Room#2", "Friday"}] = 25; // CT-24312 to CT-24318 (partial)
    // Saturday - AI401
    data.roomLoadPerDay[{"Room#1", "Saturday"}] = 55; // CT-24001 to CT-24085
    data.roomLoadPerDay[{"Room#2", "Saturday"}] = 55; // CT-24086 to CT-24166
    data.roomLoadPerDay[{"Room#3", "Saturday"}] = 55; // CT-24169 to CT-24249
    data.roomLoadPerDay[{"Room#4", "Saturday"}] = 47; // CT-24252 to CT-24320
    // Sunday - CS301
    data.roomLoadPerDay[{"Room#1", "Sunday"}] = 55; // CT-24002 to CT-24311
    // Monday - CN601
    data.roomLoadPerDay[{"Room#2", "Monday"}] = 55; // CT-24073 to CT-24137
    data.roomLoadPerDay[{"Room#3", "Monday"}] = 55; // CT-24139 to CT-24203
    data.roomLoadPerDay[{"Room#4", "Monday"}] = 55; // CT-24205 to CT-24269
    data.roomLoadPerDay[{"Room#5", "Monday"}] = 42; // CT-24273 to CT-24320
    
    // Set unused room-day combinations to 0
    for (const auto& room : rooms) {
        for (const auto& day : days) {
            if (data.roomLoadPerDay.find({room, day}) == data.roomLoadPerDay.end()) {
                data.roomLoadPerDay[{room, day}] = 0;
            }
        }
    }
    
    // Student exam distribution based on roll number analysis
    // All students (CT-24001 to CT-24320) appear to be taking multiple exams
    data.studentExamCountDistribution[1] = 50;  // Some students with 1 exam
    data.studentExamCountDistribution[2] = 80;  // Students with 2 exams
    data.studentExamCountDistribution[3] = 120; // Students with 3 exams (most common)
    data.studentExamCountDistribution[4] = 60;  // Students with 4 exams
    data.studentExamCountDistribution[5] = 10;  // Few students with all 5 exams
    
    return data;
}

void AnalyticsWindow::createExamLoadChart()
{
    // Create bar chart for exam load overview
    auto series = new QBarSeries();
    auto barSet = new QBarSet("Exams");
    
    QStringList categories;
    
    for (const auto& pair : m_analyticsData.examCountPerDay) {
        categories << QString::fromStdString(pair.first);
        *barSet << pair.second;
    }
    
    series->append(barSet);
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ“Š Exams Scheduled per Day");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisX = new QBarCategoryAxis();
    axisX->append(categories);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    auto axisY = new QValueAxis();
    axisY->setTitleText("Number of Exams");
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    // Style the chart
    chart->legend()->setAlignment(Qt::AlignBottom);
    barSet->setColor(QColor(52, 152, 219));
    
    m_examLoadChartView = new QChartView(chart);
    m_examLoadChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->examLoadChartWidget);
    auto examLoadLayout = new QVBoxLayout();
    examLoadLayout->setContentsMargins(8,8,8,8);
    ui->examLoadChartWidget->setLayout(examLoadLayout);
    ui->examLoadChartWidget->layout()->addWidget(m_examLoadChartView);
}

void AnalyticsWindow::createDepartmentDistributionChart()
{
    // Create pie chart for department distribution
    auto series = new QPieSeries();
    
    int total = 0;
    for (const auto& pair : m_analyticsData.examCountPerDepartment) {
        total += pair.second;
    }
    
    QStringList colors = {"#e74c3c", "#3498db", "#2ecc71", "#f39c12", "#9b59b6", "#1abc9c"};
    int colorIndex = 0;
    
    for (const auto& pair : m_analyticsData.examCountPerDepartment) {
        double percentage = (static_cast<double>(pair.second) / total) * 100.0;
        auto slice = series->append(QString::fromStdString(pair.first), pair.second);
        slice->setLabelVisible(true);
        slice->setLabel(QString("%1 (%2%)").arg(QString::fromStdString(pair.first)).arg(QString::number(percentage, 'f', 1)));
        slice->setColor(QColor(colors[colorIndex % colors.size()]));
        colorIndex++;
    }
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ¥§ Exam Distribution by Department");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    chart->legend()->setAlignment(Qt::AlignRight);
    
    m_departmentChartView = new QChartView(chart);
    m_departmentChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->departmentChartWidget);
    auto deptLayout = new QVBoxLayout();
    deptLayout->setContentsMargins(8,8,8,8);
    ui->departmentChartWidget->setLayout(deptLayout);
    ui->departmentChartWidget->layout()->addWidget(m_departmentChartView);
}

void AnalyticsWindow::createRoomUtilizationChart()
{
    // Create horizontal bar chart for room utilization
    auto series = new QHorizontalBarSeries();
    auto barSet = new QBarSet("Utilization %");
    
    QStringList categories;
    
    for (const auto& pair : m_analyticsData.roomUtilizationPercentage) {
        categories << QString::fromStdString(pair.first);
        *barSet << pair.second;
    }
    
    series->append(barSet);
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ“ˆ Room Capacity Utilization");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisY = new QBarCategoryAxis();
    axisY->append(categories);
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    auto axisX = new QValueAxis();
    axisX->setTitleText("Utilization Percentage");
    axisX->setRange(0, 100);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    // Color coding based on utilization level
    barSet->setColor(QColor(46, 204, 113)); // Green for good utilization
    
    chart->legend()->setAlignment(Qt::AlignBottom);
    
    m_roomUtilizationChartView = new QChartView(chart);
    m_roomUtilizationChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->roomUtilizationChartWidget);
    auto roomUtilLayout = new QVBoxLayout();
    roomUtilLayout->setContentsMargins(8,8,8,8);
    ui->roomUtilizationChartWidget->setLayout(roomUtilLayout);
    ui->roomUtilizationChartWidget->layout()->addWidget(m_roomUtilizationChartView);
}

void AnalyticsWindow::createConflictDetectionChart()
{
    // Create bar chart for conflict detection
    auto series = new QBarSeries();
    auto barSet = new QBarSet("Conflicts");
    
    QStringList categories;
    
    for (const auto& pair : m_analyticsData.conflictsPerCourse) {
        categories << QString::fromStdString(pair.first);
        *barSet << pair.second;
    }
    
    series->append(barSet);
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("âš¡ Conflicts Detected per Course");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisX = new QBarCategoryAxis();
    axisX->append(categories);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    auto axisY = new QValueAxis();
    axisY->setTitleText("Number of Conflicts");
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    // Red color for conflicts (warning)
    barSet->setColor(QColor(231, 76, 60));
    
    chart->legend()->setAlignment(Qt::AlignBottom);
    
    m_conflictChartView = new QChartView(chart);
    m_conflictChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->conflictChartWidget);
    auto conflictLayout = new QVBoxLayout();
    conflictLayout->setContentsMargins(8,8,8,8);
    ui->conflictChartWidget->setLayout(conflictLayout);
    ui->conflictChartWidget->layout()->addWidget(m_conflictChartView);
}

void AnalyticsWindow::createExamTimingChart()
{
    // Create a Gantt-style chart using scatter plot with different markers for time visualization
    auto series = new QScatterSeries();
    series->setName("Exam Timings");
    series->setMarkerSize(15.0);
    
    // Sample time slots data (would come from actual exam schedule)
    QStringList timeSlots = {"09:00-12:00", "14:00-17:00", "09:00-11:00", "11:30-14:30", "15:00-18:00"};
    QStringList courses = {"ML101", "AI401", "DSA201", "DB301", "OS401"};
    
    for (int i = 0; i < timeSlots.size() && i < courses.size(); ++i) {
        series->append(i, i); // Simple positioning for demo
    }
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ“Š Exam Timings Overview");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisX = new QValueAxis();
    axisX->setTitleText("Time Period");
    axisX->setRange(0, 6);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    auto axisY = new QValueAxis();
    axisY->setTitleText("Courses");
    axisY->setRange(0, 6);
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    series->setColor(QColor(155, 89, 182));
    chart->legend()->setAlignment(Qt::AlignBottom);
    
    m_examTimingChartView = new QChartView(chart);
    m_examTimingChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->examTimingChartWidget);
    auto timingLayout = new QVBoxLayout();
    timingLayout->setContentsMargins(8,8,8,8);
    ui->examTimingChartWidget->setLayout(timingLayout);
    ui->examTimingChartWidget->layout()->addWidget(m_examTimingChartView);
}

void AnalyticsWindow::createRoomLoadOverDaysChart()
{
#ifdef HAVE_DATAVISUALIZATION
    // Create 3D surface chart for room load over days
    m_roomLoad3DSurface = new Q3DSurface();
    
    // Create surface series
    auto surfaceSeries = new QSurface3DSeries();
    surfaceSeries->setName("Room Load");
    
    // Create sample data array
    auto dataArray = new QSurfaceDataArray();
    
    std::vector<std::string> rooms = {"Room#1", "Room#2", "Room#3", "Room#4", "Room#5"};
    std::vector<std::string> days = {"Friday", "Saturday", "Sunday", "Monday", "Tuesday"};
    
    dataArray->reserve(rooms.size());
    for (int room = 0; room < rooms.size(); ++room) {
        auto newRow = new QSurfaceDataRow(days.size());
        for (int day = 0; day < days.size(); ++day) {
            auto it = m_analyticsData.roomLoadPerDay.find({rooms[room], days[day]});
            float load = (it != m_analyticsData.roomLoadPerDay.end()) ? it->second : 0;
            (*newRow)[day].setPosition(QVector3D(day, load, room));
        }
        dataArray->append(newRow);
    }
    
    surfaceSeries->dataProxy()->resetArray(dataArray);
    m_roomLoad3DSurface->addSeries(surfaceSeries);
    
    // Configure axes
    m_roomLoad3DSurface->axisX()->setTitle("Days");
    m_roomLoad3DSurface->axisY()->setTitle("Student Load");
    m_roomLoad3DSurface->axisZ()->setTitle("Rooms");
    
    // Set axis ranges
    m_roomLoad3DSurface->axisX()->setRange(0, days.size() - 1);
    m_roomLoad3DSurface->axisY()->setRange(0, 100);
    m_roomLoad3DSurface->axisZ()->setRange(0, rooms.size() - 1);
    
    // Create widget container for 3D surface
    m_roomLoad3DWidget = QWidget::createWindowContainer(m_roomLoad3DSurface);
    m_roomLoad3DWidget->setMinimumSize(500, 280);
#else
    // Fallback to 2D chart when DataVisualization is not available
    auto series = new QLineSeries();
    series->setName("Average Room Load");
    
    std::vector<std::string> days = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday"};
    
    for (int i = 0; i < days.size(); ++i) {
        int totalLoad = 0;
        int roomCount = 0;
        
        for (const auto& pair : m_analyticsData.roomLoadPerDay) {
            if (pair.first.second == days[i]) {
                totalLoad += pair.second;
                roomCount++;
            }
        }
        
        double averageLoad = roomCount > 0 ? (double)totalLoad / roomCount : 0;
        series->append(i, averageLoad);
    }
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ“ˆ Average Room Load Over Days");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisX = new QValueAxis();
    axisX->setTitleText("Days");
    axisX->setRange(0, days.size() - 1);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    auto axisY = new QValueAxis();
    axisY->setTitleText("Average Load");
    axisY->setRange(0, 100);
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    series->setColor(QColor(52, 152, 219));
    chart->legend()->setAlignment(Qt::AlignBottom);
    
    auto chartView = new QChartView(chart);
    chartView->setRenderHint(QPainter::Antialiasing);
    
    m_roomLoad3DWidget = chartView;
#endif
    
    // Add to layout
    clearWidgetLayout(ui->roomLoadChartWidget);
    auto roomLoadLayout = new QVBoxLayout();
    roomLoadLayout->setContentsMargins(8,8,8,8);
    ui->roomLoadChartWidget->setLayout(roomLoadLayout);
    ui->roomLoadChartWidget->layout()->addWidget(m_roomLoad3DWidget);
}

void AnalyticsWindow::createStudentExamDistributionChart()
{
    // Create histogram for student exam count distribution
    auto series = new QBarSeries();
    auto barSet = new QBarSet("Students");
    
    QStringList categories;
    
    for (const auto& pair : m_analyticsData.studentExamCountDistribution) {
        categories << QString("%1 Exams").arg(pair.first);
        *barSet << pair.second;
    }
    
    series->append(barSet);
    
    auto chart = new QChart();
    chart->addSeries(series);
    chart->setTitle("ðŸ‘¨â€ðŸŽ“ Number of Exams per Student");
    chart->setAnimationOptions(QChart::SeriesAnimations);
    
    auto axisX = new QBarCategoryAxis();
    axisX->append(categories);
    chart->addAxis(axisX, Qt::AlignBottom);
    series->attachAxis(axisX);
    
    auto axisY = new QValueAxis();
    axisY->setTitleText("Number of Students");
    chart->addAxis(axisY, Qt::AlignLeft);
    series->attachAxis(axisY);
    
    // Gradient color for better visualization
    barSet->setColor(QColor(26, 188, 156));
    
    chart->legend()->setAlignment(Qt::AlignBottom);
    
    m_studentDistributionChartView = new QChartView(chart);
    m_studentDistributionChartView->setRenderHint(QPainter::Antialiasing);
    
    // Add to layout
    clearWidgetLayout(ui->studentDistributionChartWidget);
    auto studentLayout = new QVBoxLayout();
    studentLayout->setContentsMargins(8,8,8,8);
    ui->studentDistributionChartWidget->setLayout(studentLayout);
    ui->studentDistributionChartWidget->layout()->addWidget(m_studentDistributionChartView);
}

// Slot implementations
void AnalyticsWindow::onRefreshClicked()
{
    refreshAnalytics();
    QMessageBox::information(this, "Analytics Refreshed", "Analytics data has been refreshed successfully!");
}

void AnalyticsWindow::onExportClicked()
{
    QString fileName = QFileDialog::getSaveFileName(this, "Export Analytics Report", 
                                                   "analytics_report.pdf", "PDF Files (*.pdf)");
    if (!fileName.isEmpty()) {
        exportChartsAsPDF(fileName);
        QMessageBox::information(this, "Export Successful", 
                                QString("Analytics report exported to: %1").arg(fileName));
    }
}

void AnalyticsWindow::onCloseClicked()
{
    // Accept the dialog to close it properly
    accept();
}

void AnalyticsWindow::keyPressEvent(QKeyEvent* event)
{
    if (event->key() == Qt::Key_Escape) {
        accept(); // Close dialog on ESC
    } else {
        QDialog::keyPressEvent(event);
    }
}

void AnalyticsWindow::closeEvent(QCloseEvent* event)
{
    // Accept the close event
    event->accept();
}

// Data processing methods (stubs for database integration)
void AnalyticsWindow::processExamLoadData()
{
    // Implementation would fetch and process real exam schedule data
    // For now, using sample data
    generateSampleData();
}

void AnalyticsWindow::processDepartmentDistributionData()
{
    // Implementation would analyze course-department relationships
}

void AnalyticsWindow::processRoomUtilizationData()
{
    // Implementation would calculate room capacity vs actual usage
}

void AnalyticsWindow::processConflictDetectionData()
{
    // Implementation would detect scheduling conflicts
}

void AnalyticsWindow::processExamTimingData()
{
    // Implementation would analyze exam timing overlaps
}

void AnalyticsWindow::processRoomLoadOverDaysData()
{
    // Implementation would track room usage over time
}

void AnalyticsWindow::processStudentExamDistributionData()
{
    // Implementation would count exams per student
}

std::vector<ExamScheduleData> AnalyticsWindow::getExamScheduleFromDatabase()
{
    std::vector<ExamScheduleData> schedule;
    
    if (m_databaseManager) {
        // Implementation would fetch actual schedule data from database
        // For now, return empty to use sample data
    }
    
    return schedule;
}

void AnalyticsWindow::exportChartsAsPDF(const QString& filename)
{
    QPrinter printer(QPrinter::PrinterResolution);
    printer.setOutputFormat(QPrinter::PdfFormat);
    printer.setOutputFileName(filename);
    printer.setPageSize(QPageSize::A4);
    printer.setPageMargins(QMarginsF(15, 15, 15, 15), QPageLayout::Millimeter);
    
    QPainter painter;
    if (painter.begin(&printer)) {
        // Capture the entire analytics window
        QPixmap pixmap = this->grab();
        
        QRect rect = painter.viewport();
        QSize size = pixmap.size();
        size.scale(rect.size(), Qt::KeepAspectRatio);
        
        painter.setViewport(rect.x(), rect.y(), size.width(), size.height());
        painter.setWindow(pixmap.rect());
        painter.drawPixmap(0, 0, pixmap);
        
        painter.end();
    }
}

QColor AnalyticsWindow::getRandomColor(int index)
{
    QStringList colors = {"#e74c3c", "#3498db", "#2ecc71", "#f39c12", "#9b59b6", "#1abc9c", "#34495e", "#e67e22"};
    return QColor(colors[index % colors.size()]);
}

QString AnalyticsWindow::formatPercentage(double value)
{
    return QString::number(value, 'f', 1) + "%";
}